tools/richlist.js

```diff
@@ -0,0 +1,564 @@
+#!/usr/bin/env node
+/**
+ * Tool for calculating richlist by hackyminer
+ */
+
+var _ = require('lodash');
+var Web3 = require('web3');
+var async = require('async');
+var BigNumber = require('bignumber.js');
+var mongoose = require('mongoose');
+
+var Account = require('../db.js').Account;
+var Transaction = require('../db.js').Transaction;
+var Block = require('../db.js').Block;
+
+mongoose.connect(process.env.MONGO_URI || 'mongodb://localhost/blockDB');
+
+const ADDRESS_CACHE_MAX = 10000; // address cache threshold
+
+// RichList for Geth Classic, Geth
+function makeRichList(toBlock, blocks, updateCallback) {
+  var self = makeRichList;
+  if (!self.cached) {
+    self.cached = {};
+    self.index = 0;
+  }
+  if (!self.accounts) {
+    self.accounts = {};
+  }
+  var fromBlock = toBlock - blocks;
+  if (fromBlock < 0) {
+    fromBlock = 0;
+  }
+
+  console.log('Scan accounts from ' + fromBlock + ' to ' + toBlock + ' ...');
+
+  var ended = false;
+  if (fromBlock == toBlock) {
+    ended = true;
+  }
+
+  async.waterfall([
+    function(callback) {
+      // Transaction.distinct("from", { blockNumber: { $lte: toBlock, $gt: fromBlock } }, function(err, docs) ...
+      // faster
+      // dictint("from")
+      Transaction.aggregate([
+        { $match: { blockNumber: { $lte: toBlock, $gt: fromBlock } } },
+        { $group: { _id: '$from' }},
+        { $project: { "_id": 1 }}
+      ]).exec(function(err, docs) {
```

Comment:
I think you can abstract this part into a separate function you can pass into `exec`, since it's being repeated in the next three code chunks.

Reply:
Ok, since you said you'll remedy it in a separate PR after this one is merged, I'll resolve this for now.